crates/wasmparser/src/readers/core/types.rs

```diff
@@ -50,6 +53,12 @@ const _: () = {
     assert!(std::mem::size_of::<ValType>() == 4);
 };
 
+pub(crate) trait Matches {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>;
```

Comment:
One thing I might recommend is to remove the `Result` here since indices at this point should always be valid because the definitions of types are validated to all be correct. Additionally it might be better to pass a concrete type here rather than a generic `Fn(u32) -> &SubType` but instead have something like `cx: &Validator` or similar.

Reply:
> One thing I might recommend is to remove the Result here since indices at this point should always be valid because the definitions of types are validated to all be correct.

Actually, the primary use case for `matches()` is to validate if the type definitions are correct.

Reply:
I don't think that's the case? Whenever a type definition is reached as part of validation there are no type-matching queries performed at that time and instead the structure of the defined type is validated. This means that the indices are all valid for the type section after it's read, so in the future when the matching predicate comes into play during the code section for example there are no invalid indices.

I believe this is `Module::add_type` which is where indices are validated and afterwards indices don't need validation.

Reply:
> I believe this is Module::add_type which is where indices are validated and afterwards indices don't need validation.

Yeah, that's exactly where `matches()` gets called:

```
<wasmparser::readers::core::types::SubType as wasmparser::readers::core::types::Matches>::matches types.rs:763
wasmparser::validator::core::Module::check_subtype core.rs:525
wasmparser::validator::core::Module::add_type core.rs:507
```

Reply:
Ah ok I see that now, but I still think that this `Result` isn't necessary. For example if `check_structural_type` is moved above the subtype check, then everything is guaranteed to be valid while performing a subtype check?

Reply:
Thanks! Updated.